### DISCUSSION TITLE
Why use Struts?

### DISCUSSION THREAD

**Discussion by Andywmm9** [nguyenquyhy/Flight-Tracker-StreamDeck#153] [Ideas]
When we want to add a new SimConnect variable, Structs.cs has to be modified.  Can I ask why we just can't treat SimConnect variables as a string?  That way, when a new variable comes along, the program won't have to be modified.

**Comment by rmroc451** [comment]
It's a sort of gateway to make sure things from the API are working.

If we don't implement them because they don't work, that prevents folks from trying to use them and getting frustrated when things are broken or not working as expected 👍

**Comment by nguyenquyhy** [comment]
Basically we don't want to limit users to a fixed list, but at the same time, we don't want to lose the benefits coming from a pre-defined list:
- It is relatively hard to handle error properly in SimConnect. If you have a button with wrong variable name, SimConnect gives you an NAME_UNRECOGNIZED error without much information to know which button is having the issue. 
  - A little bit deeper on some technical details about this: SimConnect allows us to send requests to the sim with a list of variables and it will later response with a data packet containing the values of those variables. To make sure we don't overload SimConnect, we bundle all generic variables in all displaying buttons and send them in 1 request. However, if one of the variable names is incorrect, SimConnect returns a data packets with 1 fewer value, and we have no way to know which value is missing. I know many (or all?) other add-ons do this differently: they send each variable in a request, which allows them to ignore the error because SimConnect won't returns anything for the request.
  - Knowing how unstable and unpredictable SimConnect is (not just in MSFS but also in FSX and P3D), I try to avoid causing error as much as possible.
- We have some plans on auto-completion, which needs the support of a pre-defined list. Variable unit is another place our users can put wrongly, and our current list automatically fill that in for them.

But as I said in the beginning, it's not really our aim to restrict our users but to try our best to provide as much stability and user friendliness as possible. We currently have a PR from RMROC for custom units, a branch from Pieter for variable dropdown, and I have some ideas in mind on opening up for any events/variables without scarifying our SimConnect optimization. However, I am taking a bit more time to think this through to make sure we can deliver the following features:
- Allow entering any variables and units without causing a flood of errors from SimConnect
- Auto-completion for known variables/events
- Compatible with all current profiles and still easy enough to use

And I think I am almost there 😛.

P/S: It took me too long to write this and RMROC beats me to it 😂.

**Comment by rkantos** [comment]
Uuntil someone figures out a proper way to handle spamming any SimConnect requests as fast as you can push the StreamDeck buttons, the issues page would just be filled with more of these:

#152 #56 #55 #34 #24

**Reply by nguyenquyhy** [reply]
I don't think the loop is due to spamming requests. I hold and spam the button a lot to change heading and altitude, and I have almost never got the loop after Alpha. It does cause an UNKNOWN ERROR sometimes, which leads to an intentional reconnect from the plugin, but it is fine after the reconnection without the death loop. But well, it's SimConnect, who knows 😂?

**Reply by rmroc451** [reply]
There have been some posts in the forums about issues with the API when it comes underload.  I wonder if this is the underlying issue and will be fixed with whatever is also causing the CTDs that they noted were also potentially related to API calls.

**Reply by rkantos** [reply]
> I don't think the loop is due to spamming requests. I hold and spam the button a lot to change heading and altitude, and I have almost never got the loop after Alpha. It does cause an UNKNOWN ERROR sometimes, which leads to an intentional reconnect from the plugin, but it is fine after the reconnection without the death loop. But well, it's SimConnect, who knows 😂?

By spamming *any* requests I was referring to spamming invalid stuff too. I understand SimConnect still doesn't want anything to do with them..

**Reply by nguyenquyhy** [reply]
But you can't spam invalid stuff from the Stream Deck plugin because we have the whitelist, although the invalid stuffs you are mentioning about may come from other SimConnect plugins. Maybe that's why I cannot reproduce the loop; I use very few SimConnect plugins.
